pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/BaseRecordExtractor.java

```diff
@@ -0,0 +1,190 @@
+   * Returns {@code null} if the value is an empty array/collection/map.
+   *
+   * Natively Pinot only understands single values and multi values.
+   * Map is useful only if some ingestion transform functions operates on it in the transformation layer.
+   */
+  @Nullable
+  public Object convert(Object value) {
+    Object convertedValue;
+    if (isMultiValue(value)) {
+      convertedValue = convertMultiValue(value);
+    } else if (isMap(value)) {
+      convertedValue = convertMap(value);
+    } else if (isRecord(value)) {
+      convertedValue = convertRecord(value);
+    } else {
+      convertedValue = convertSingleValue(value);
+    }
+    return convertedValue;
+  }
+
+  /**
+   * Returns whether the object is of the data format's base type. Override this method if the extractor
+   * can handle the conversion of nested record types.
+   */
+  protected boolean isRecord(Object value) {
+    return false;
+  }
+
+  /**
+   * Returns whether the object is of a multi-value type. Override this method if the data format represents
+   * multi-value objects differently.
+   */
+  protected boolean isMultiValue(Object value) {
+    return value instanceof Collection;
+  }
+
+  /**
+   * Returns whether the object is of a map type. Override this method if the data format represents map objects
+   * differently.
+   */
+  protected boolean isMap(Object value) {
+    return value instanceof Map;
+  }
+
+  /**
+   * Handles the conversion of every field of the object for the particular data format. Override this method if the
+   * extractor can convert nested record types.
+   *
+   * @param value should be verified to be a record type prior to calling this method as it will be handled with this
+   *              assumption
+   */
+  @Nullable
+  protected Object convertRecord(Object value) {
+    throw new UnsupportedOperationException("Extractor cannot convert record type structures for this data format.");
+  }
+
+  /**
+   * Handles the conversion of each element of a multi-value object. Returns {@code null} if the field value is
+   * {@code null}.
+   *
+   * This implementation converts the Collection to an Object array. Override this method if the data format
+   * requires a different conversion for its multi-value objects.
+   *
+   * @param value should be verified to be a Collection type prior to calling this method as it will be casted
+   *              to a Collection without checking
+   */
+  @Nullable
+  protected Object convertMultiValue(Object value) {
+    Collection collection = (Collection) value;
+    if (collection.isEmpty()) {
+      return null;
+    }
+
+    int numValues = collection.size();
+    Object[] array = new Object[numValues];
+    int index = 0;
+    for (Object element : collection) {
+      Object convertedValue = null;
+      if (element != null) {
+        convertedValue = convert(element);
+      }
+      if (convertedValue != null && !convertedValue.toString().equals("")) {
```

Comment:
Shall we add some comments here about the behavior of the empty string?

Reply:
I'll mention this nuance in the method's javadoc.